### ISSUE TITLE
Get element attribute via map-like keyword lookup

### ISSUE THREAD

**Issue by bsima** [hoplon/hoplon#76]
Originally [proposed on slack](https://clojurians.slack.com/archives/hoplon/p1447334426000029) by @thedavidmeister. I'm attaching a screenshot in case slack deletes our archive :imp: 

![snapshot1](https://cloud.githubusercontent.com/assets/200617/11120520/4dcfc606-891e-11e5-9be3-2674102960f6.png)

I think all this would take is implementing [IMapEntry](https://github.com/clojure/clojurescript/blob/master/src/main/cljs/cljs/core.cljs#L527-L532) on [js/Element here](https://github.com/hoplon/hoplon/blob/master/src/hoplon/core.cljs#L263-L297).

Then you could do something like the following:

```clojure
(def el (a :href "http://www.example.com" "Link to example"))

(:href el) ;;#=> "www.example.com"
```

**Comment by pbzdyl** [comment]
I think you would need to extend `js/Element` to [`ILookup`](https://github.com/clojure/clojurescript/blob/master/src/main/cljs/cljs/core.cljs#L506). I have a pull request ready, let's see what @micha and @alandipert have to say about it.

**Comment by bsima** [comment]
I view this as more of a syntactic extension than anything:

```clojure
(def my-elem (a :href x :class (cell= "whatever")))
```

So I assume `(:class my-elem)` returns the cell? Because that seems like it would be useful.

**Comment by alandipert** [comment]
@bsima that's the direction we want to go, which is why I think tying ILookup/get to jQuery's .attr is premature, if awesome.  @micha could correct me.

**Comment by micha** [comment]
We used to have this at the very beginning, but it was never very useful so it was removed pretty quickly. The reason why we never found it to be useful is because we don't generally need to store information in DOM elements. The information is in cells and vars, and usually that's where we want to refer to them. When we do want to interact with some DOM elements that were perhaps created externally or something like that we can use jQuery `.attr()` or Element `.getAttribute()`.

I'm curious what the use case is? I've always found in the past that whenever I wanted this I always ended up realizing that there is a better way that didn't involve storing data in the DOM.

**Comment by pbzdyl** [comment]
I am closing the pull request then.

**Comment by micha** [comment]
Sorry, I didn't mean for you to close the pull request :) Maybe we do want to store some information in the DOM?

**Comment by pbzdyl** [comment]
I just see we need more discussion before posting the solution :). For example I realised that maybe we shouldn't assume the key will be always a keyword but maybe any kind of object that would need to be "stringified" somehow. Once we have a consensus if we want and how we want it implemented I wil be happy to open a new pull request.

**Comment by bsima** [comment]
I don't think it's a good idea to store info in the DOM either, but in a full-fledged SPA that wouldn't work anyway, so the programmer would quickly realize that there's a better way. 

Looking at this proposal as a syntactic extension is interesting to me because it allows you to use elements like other Clojure data structures

```clojure
(def a-map {:a "a" :b "b"})
(:a a-map) ;;#=> "a"

(def a-link (a :href "a" :class "b"))
(:href a-link) ;;#=> "a"
```

I don't use Hoplon on a daily basis like you other guys do, but I'm thinking that treating elements like regular Clojure data structures can be useful, for example if you're trying to build an element programatically...

**Comment by thedavidmeister** [comment]
I originally asked for this. My use case at the time was wanting to map over a vector of links and do something to each based on the class on each.

I had actually assumed that elements would work like a collection in some way, but they don't, so this violated the principle of least surprise and left me clutching for jQuery methods to get that the classes on my elements.

**Comment by thedavidmeister** [comment]
also, can I just point out that it's literally impossible to store zero information in the DOM, so can we just not even go there please?

**Comment by thedavidmeister** [comment]
Here's an example:

````
; If the href of an element is the current path add data-current-path
(defmethod c/do! :data-current-path
  [elem _ _]
  (if (url/current-path? ($/attr elem "href"))
    (elem :attr {"data-current-path" ""})))
````

Where you can probably imagine what `url/current-path?` does.

I think I'd prefer to do:

````
; If the href of an element is the current path add data-current-path
(defmethod c/do! :data-current-path
  [elem _ _]
  (if (url/current-path? (:href elem))
    (elem :attr {"data-current-path" ""})))
````

**Comment by thedavidmeister** [comment]
So, we discussed this more in chat. Part of the chat was specific to what I was trying to do, part of it was more generic and relevant to this thread.

The synopsis (apologies if I've misread/represented anything), without really coming to any conclusion:

- @onetom pointed out that having `(:foo elem)` syntax might encourage people to break the one-way data flow philosophy by getting some attribute then setting that same attribute again, or just plain old mixing data/presentation by getting an attribute from the DOM as a way control the program or process other data. Everyone agreed that doing this is a bad idea. I don't know whether we came to consensus on what extent we think the syntax sugar for getting attributes would really encourage/discourage these ideas from becoming reality.
- @micha pointed out that for many custom element types `(:foo elem)` wouldn't exist because `:foo` could do a bunch of things that don't result in a key-value style store within the element at all, even if we wanted to get at it later, we can't - although despite this, apparently earlier versions of Hoplon did have a way to support/handle this problem (I assume it was "grab attributes if they are there, fail for everything else")
- I pointed out that being able to use clojure's functions that work with collections in order to aid with programming your *presentation* (If you mentally replace my `data-current-path` attribute above with inline CSS it's hopefully clearer where I'm coming from) seems like a legitimate enough goal to me (shrug)
